### PR TITLE
Generate null check code in existing methods for special bridge methods.

### PR DESCRIPTION
**Pull request by madsager**

### CODE REVIEW

**Comment by madsager** [comment]
Friendly ping. :)

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/BridgeLowering.kt [hunk 1]
```diff
@@ -158,6 +157,48 @@ private class BridgeLowering(val context: JvmBackendContext) : ClassLoweringPass
...
+                // Change the first parameter type to be nullable so that null checks are not generated. The checks
+                // we insert here make them superfluous.
+                irFunction.valueParameters[0] = firstParameter.copyTo(irFunction, type = firstParameterType.makeNullable())
+                // Insert an explicit null check and a return of the default value directly in the method.
+                val bodyStartOffset = irFunction.body!!.startOffset
+                val bodyEndOffset = irFunction.body!!.endOffset
+                irFunction.body = IrBlockBodyImpl(bodyStartOffset, bodyEndOffset).apply {
+                    statements += IrWhenImpl(bodyStartOffset, bodyStartOffset, context.irBuiltIns.unitType).apply {
+                        val nullConst = IrConstImpl.constNull(bodyStartOffset, bodyStartOffset, context.irBuiltIns.booleanType)
+                        val equals =
```
Comment (gogabr) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/BridgeLowering.kt]: Corresponding check in `createBridgeBody` is more readable -- why not copy it here, or even create an auxiliary function?

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/BridgeLowering.kt [hunk 2]
```diff
@@ -158,6 +157,48 @@ private class BridgeLowering(val context: JvmBackendContext) : ClassLoweringPass
             )
         }
 
+        // If there is an existing function that would conflict with a special bridge signature, insert the special bridge
+        // code directly as a prelude in the existing method.
+        if (!irFunction.isFakeOverride && specialOverride != null && specialOverrideSignature == ourSignature) {
+            val firstParameter = irFunction.valueParameters[0]
+            val firstParameterType = firstParameter.type
+            if (!firstParameterType.isNullable()) {
```
[the remark this answers is not in the document]
Comment (gogabr) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/BridgeLowering.kt]: Another point: the check in `createBridgeBody` is not a null check but an `isinstance` check.
It looks like this is more appropriate here as well (changing the parameter type to `Any?`).

[the remark this answers is not in the document]
Comment (madsager) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/BridgeLowering.kt]: Thanks! The correct fix is to be more explicit about the number of arguments to check. For `getOrDefault` we need to check only the first argument. For `remove(key, value)` we need to check both.

Changed the null check to an `isinstance` check and the type to `Any?`.

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/BridgeLowering.kt (hunk 1 above)
Comment (madsager) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/BridgeLowering.kt]: Thank you!!! Yes, that is more readable (and writable). Changed.

**Comment by madsager** [comment]
Rebased after test file changes. Ready for another look. :)

**Comment by gogabr** [comment]
Thank you!

**Pull request merged**